frontend/app/templates/index.html

```diff
@@ -0,0 +1,507 @@
+                                let property = schema.properties[prop];
+                                if (isFilterable(property)) {
+                                    // merge in `multiple` to keep track of the type of containment
+                                    // (key + # + prop is necessary to namespace the filter ids)
+                                    schemaFilters[schema.title][key + '#' + prop] = Object.assign({}, property, {multiple: schema.multiple}); }
+                            });
+                        });
+
+                        Object.keys(schemaFilters).forEach((schemaName) => {
+                            // Create a new accordion to contain filters for this schema.
+                            $('#extra-filter-container').append(`
+                                <div class="col-xs-12" style="margin-top:5px;">
+                                    <h5>Filter by ${schemaName}</h5>
+                                    <hr />
+                                    <div class="accordion" name="${schemaName}"></div>
+                                </div>
+                            `);
+
+                            let filterables = schemaFilters[schemaName];
+
+                            Object.keys(filterables).forEach((filterName) => {
+                                let filterable = filterables[filterName];
+
+                                let widget = getFilterWidget(filterName, filterable);
+
+                                $(`.accordion[name="${schemaName}"]`).append(widget);
+                            });
+                        });
+
+                  }).fail((err) => {
+                        console.log(err);
+                  });
+            }
+
+            function getFilterWidget(key, filterable) {
+                // Given a `filterable` object and its name `key`,
+                // return an HTML element for the filter widget.
+                let num = filterable.propertyOrder;
+                let label = key.split('#')[1];  // Extract the field name from the namespaced key
+                let elem = `
+                    <div class="card">
+                        <div class="card-header">
+                            <h5 class="mb-0">
+                                <button class="btn btn-link" type="button"
+                                        data-toggle="collapse"
+                                        data-target="#filter-dropdown-${num}">
+                                    ${label}
+                                </button>
+                            </h5>
+                        </div>
+                        <div id="filter-dropdown-${num}" class="collapse">
+                            <div class="card-body">
+                `;
+
+                switch (filterable.fieldType) {
+                    case 'number':
+                        // Number types should produce range dropdowns.
+                        elem += `
```

Comment:
DRIVER had really nice abstractions for these components, I thought; pulling them out into this function kind of felt like breaking down a nicely-architected piece of code back into a hack. I'll probably take inspiration from DRIVER when I redo this work in a component-oriented setting.

Related concern: in the abstract it seems like it'd be a nice feature for the JS library to produce some of these components for the developer, since handling all the different types of filters is a pain. But is there any way of approaching that goal in a framework-agnostic fashion?

Reply:
Hmmm, my initial reaction is that generating components would be the kind of thing that would be good for framework-specific libraries to handle. But maybe there's still some useful work that a JS library could do to help with parsing the schema so that that part doesn't have to be repeated.